examples/cpp/pyperf/PyPerfNativeStackTrace.cc

```diff
@@ -65,23 +81,33 @@ NativeStackTrace::NativeStackTrace(uint32_t pid, const unsigned char *raw_stack,
     goto out;
   }
 
+  struct bcc_symbol resolved_symbol;
+  if (NativeStackTrace::insert_dso_name) {
+    procSymbols = get_proc_symbols(pid);
+  }
+
   do {
     unw_word_t offset;
-    char sym[256];
-    char   *realname;
-    int     status;
-
     // TODO: This function is very heavy. We should try to do some caching here, maybe in the
     //       underlying UPT function.
     res = unw_get_proc_name(&cursor, sym, sizeof(sym), &offset);
     if (res == 0) {
-      realname = abi::__cxa_demangle(sym, NULL, NULL, &status);
-      if (status == 0) {
-        this->symbols.push_back(std::string(realname));
+      unw_word_t ip;
+      unw_get_reg(&cursor, UNW_REG_IP, &ip);
+      if (procSymbols && procSymbols->resolve_addr(ip, &resolved_symbol, true)) {
+        snprintf(buf.get(), buf_size, "%s (%s)", resolved_symbol.demangle_name, resolved_symbol.module);
+        this->symbols.push_back(std::string(buf.get()));
       } else {
-        this->symbols.push_back(std::string(sym));
+        int status = 0;
+        char* demangled = nullptr;
+        demangled = abi::__cxa_demangle(sym, nullptr, nullptr, &status);
+        if (!status) {
+          this->symbols.push_back(std::string(demangled));
+          free(demangled);
+        } else {
+          this->symbols.push_back(std::string(sym));
+        }
       }
-      free(realname);
```

Comment:
Previously we `free`d in all cases, even if `__cxa_demangle` reportedly fails. Now you `free` only in the success path? Any reason for this change?

I don't recall why we originally did it this way - please see in the PR why, explain the reason to change or revert the change such that `free` is always called in the `__cxa_demangle` path.

Reply:
This change wasn't necessary. I guess I lacked the sense of safety seeing `free()` called when argument might be `NULL`.
I will revert it to the previous state, in the vein of keeping least diff necessary.

Actually that wouldn't be a mistake according to documentation. `__cxa_demangle` would only return actual buffer in the success path. Any other case, and it returns `nullptr`.

Reply:
If that's the documentation, it's fine. `free(NULL)` is always legit. I wanted to see if there was a reason for it.

Reply:
Resolving this